deepinv/transform/base.py

```diff
@@ -18,13 +85,99 @@ def __init__(self, *args, n_trans: int = 1, rng: torch.Generator = None, **kwarg
+
+    def transform(self, x: torch.Tensor, **params) -> torch.Tensor:
+        """Transform image given transform parameters.
+
+        Given randomly generated params (e.g. rotation degrees), deterministically transform the image x.
+
+        Override this to implement a custom transform.
 
         :param torch.Tensor x: input image of shape (B,C,H,W)
+        :param **params: params e.g. degrees or shifts provided as keyword args.
+        :return: torch.Tensor: transformed image.
         """
         return NotImplementedError()
 
+    def forward(self, x: torch.Tensor, **params) -> torch.Tensor:
+        """Perform random transformation on image.
+
+        Calls ``get_params`` to generate random params for image, then ``transform`` to deterministically transform.
+
+        For purely deterministic transformation, pass in custom params and ``get_params`` will be ignored.
+
+        :param torch.Tensor x: input image of shape (B,C,H,W)
+        :return torch.Tensor: randomly transformed images concatenated along the first dimension
+        """
+        return self.transform(x, **(self.get_params(x) if not params else params))
+
+    def inverse(self, x: torch.Tensor, **params) -> torch.Tensor:
+        """Perform random inverse transformation on image (i.e. when not a group).
+
+        For purely deterministic transformation, pass in custom params and ``get_params`` will be ignored.
+
+        :param torch.Tensor x: input image
+        :return torch.Tensor: randomly transformed images
+        """
+        return self.transform(
+            x, **self.invert_params(self.get_params(x) if not params else params)
+        )
+
+    def identity(self, x: torch.Tensor) -> torch.Tensor:
+        """Sanity check function that should do nothing.
+
+        This performs forward and inverse transform, which results in the exact original, down to interpolation effects.
+
+        Interpolation effects will be visible in non-pixelwise transformations, such as arbitrary rotation, scale or projective transformation.
+
+        :param torch.Tensor x: input image
+        :return torch.Tensor: :math:`T_g^{-1}T_g x=x`
+        """
+        return self.symmetrize(f=lambda _x: _x)(x)
+
+    def symmetrize(
+        self, f: Callable[[torch.Tensor, Any], torch.Tensor]
+    ) -> Callable[[torch.Tensor, Any], torch.Tensor]:
+        """Symmetrise a function with a transform and its inverse.
+
+        Given a function :math:`f(\cdot):X\rightarrow X` and a transform :math:`T_g`, return the function :math:`T_g^{-1} f(T_g \cdot)`
+
+        This is useful for e.g. Reynolds averaging a function over a group.
+
+        :param Callable[[torch.Tensor, Any], torch.Tensor] f: function acting on tensors.
+        :return Callable[[torch.Tensor, Any], torch.Tensor]: decorated function.
+        """
+
+        def symmetrized(x, *args, **kwargs):
+            params = self.get_params(x)
+            return self.inverse(
```

Comment:
We should add a parameter `n_samples=1` (or similar) which lets you average over a larger number of random monte carlo evaluations